### PR TITLE
Monitor Snyk projects on release

### LINKED ISSUE #9245: Use Snyk to check for license issues or dependency vulnerabilities
**Description**

We should integrate Snyk into our CI to check for dependency vulnerabilities and licensing issues.

For licensing issues, refer to https://confluence.camunda.com/pages/viewpage.action?pageId=76485290 to create an appropriate policy.

For vulnerabilities, we want a check which fails when introducing vulnerabilities with `Medium` or higher vulnerabilities. To allow for exceptions when the vulnerability does not affect us, you can use a `.snyk` file which you can create/edit with the Snyk CLI.

We can use the [Snyk GitHub actions](https://github.com/snyk/actions) to help us out here, as they will be more flexible than the basic GitHub integration.

### PR DESCRIPTION
**Pull request by npepinpe**
## Description

This PR adds a job to the release workflow to also monitor Snyk project for the given release. It uses the released Docker image and the local Maven POMs from the release branch, overriding the version and the Docker image name. 

During a dry run, we run a test of the Snyk artifacts instead. It's not perfect, but better than nothing, as it will at least make sure Snyk can parse our artifacts.

> **Note**
> The Docker image is built during a dry-run since no previous Docker images are built. However, the job still waits for the Docker and release jobs to go through. I'm not aware of any way to make the `needs` parameter conditional.

Projects monitored during release will be created under the grouping using only the major and minor release, e.g. 8.2, 8.1, 8.0. The project names include only the minor release as well, so projects will be replaced on patch release, meaning we only keep track of the latest release on a minor branch.

> **Warning**
> This does not contain clearing out the out-of-date minor version.

## Related issues

related to #9245

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [ ] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [ ] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Other teams:
If the change impacts another team an issue has been created for this team, explaining what they need to do to support this change.
- [ ] [Operate](https://github.com/camunda/operate/issues)
- [ ] [Tasklist](https://github.com/camunda/tasklist/issues)
- [ ] [Web Modeler](https://github.com/camunda/web-modeler/issues)
- [ ] [Desktop Modeler](https://github.com/camunda/camunda-modeler/issues)
- [ ] [Optimize](https://github.com/camunda/camunda-optimize/issues)

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

### CODE REVIEW

File: .github/workflows/snyk.yml
Comment (npepinpe) [inline .github/workflows/snyk.yml:106]: Tested locally via bash with the following combinations:

No `inputs.dockerImage`, but `inputs.build` is true.

```bash
export DOCKER_IMAGE=$([[ ! -z '' ]] && echo '' || ([[ 'true' == 'true' ]] && echo ' steps.build-docker.outputs.image' || echo "camunda/zeebe:${VERSION}")); echo $DOCKER_IMAGE
```

Prints 'steps.build-docker.outputs.image'


No `inputs.dockerImage`, and `inputs.build` is false.

```bash
export VERSION="8.1.0"
export DOCKER_IMAGE=$([[ ! -z '' ]] && echo '' || ([[ 'false' == 'true' ]] && echo ' steps.build-docker.outputs.image' || echo "camunda/zeebe:${VERSION}")); echo $DOCKER_IMAGE
```

Prints 'camunda/zeebe:8.1.0'

A given `inputs.dockerImage`, and `inputs.build` is true.

```bash
export VERSION="8.1.0"
export DOCKER_IMAGE=$([[ ! -z 'myImage' ]] && echo 'myImage' || ([[ 'true' == 'true' ]] && echo ' steps.build-docker.outputs.image' || echo "camunda/zeebe:${VERSION}")); echo $DOCKER_IMAGE
```

Prints 'myImage'


A given `inputs.dockerImage`, but `inputs.build` is false.

```bash
export VERSION="8.1.0"
export DOCKER_IMAGE=$([[ ! -z 'myImage' ]] && echo 'myImage' || ([[ 'false' == 'true' ]] && echo ' steps.build-docker.outputs.image' || echo "camunda/zeebe:${VERSION}")); echo $DOCKER_IMAGE
```

Prints 'myImage'

File: .github/workflows/release.yml
Comment (npepinpe) [inline .github/workflows/release.yml:323]: The idea is to skip alpha, snapshot, rc, etc., anything that isn't a production version. However, expressions don't support regex, so this is the simplest I came up with now. Ideas are welcome :+1:

**Comment by npepinpe** [comment]
Dry run: https://github.com/camunda/zeebe/actions/runs/5244879164/jobs/9472041928

**Comment by npepinpe** [comment]
Ah, just realized the target reference is wrong. It will always point to `release-8.2.0` or something, and that's not what we want. We want to override it here to the release version minus the patch, e.g. 8.2, 8.3, 8.1.

File: .github/workflows/snyk.yml
Comment (npepinpe) [inline .github/workflows/snyk.yml:118]: I don't like this, but I don't really know how else to grab only the minor release part of the release =/

**Comment by npepinpe** [comment]
So in the dry run I used the bash operator `ge` by mistake to specify the exit code. Corrected it to `gt` (greater than) now. But everything else looks fine as far as dry-run goes.

**Comment by npepinpe** [comment]
@megglos let me know if you don't have any time for this

**Comment by megglos** [comment]
@npepinpe apologies for the delay, I will get to it today still

File: .github/workflows/snyk.yml
Comment (megglos) [reply .github/workflows/snyk.yml:116]: I'm convinced, thanks for making the review easier :D

File: .github/workflows/snyk.yml
Comment (megglos) [inline .github/workflows/snyk.yml:119]: 🔧  LGTM, one idea that would be more strict to only catch major and minor, shouldn't matter much though 
```suggestion
            export TARGET="$(echo $VERSION | sed -E -e 's/^([0-9]+\.[0-9]+).*$/\1/')"
```

**Review by megglos — APPROVED** [review]
> The project names include only the minor release as well, so projects will be replaced on patch release, meaning we only keep track of the latest release on a minor branch.

❤️  that, will allow us to focus on the latest state easily!

LGTM, only one suggestion!

**Comment by npepinpe** [comment]
let's wait until after the retreat to merge, in case we need an emergency release in between :smile:

**Comment by npepinpe** [comment]
bors merge

**Comment by npepinpe** [comment]
@megglos - I don't think we need to backport this because the workflow always uses the definition from main, right?

**Comment by npepinpe** [comment]
bors r+

Shouldn't bors retry on merge conflicts? :roll_eyes:

**Comment by npepinpe** [comment]
Merging manually because bors is being a pain:

```erlang
{{:badmatch,
  {:error, :post_commit_status, 422,
   "{\"message\":\"Validation Failed\",\"errors\":[{\"resource\":\"Status\",\"code\":\"custom\",\"message\":\"This SHA and context has reached the maximum number of statuses.\"}],\"documentation_url\":\"https://docs.github.com/rest/commits/statuses#create-a-commit-status\"}"}}
```

**Comment by npepinpe** [comment]
It looks like we do have to backport this? The 8.2 job ran today and the Snyk job doesn't even appear.

**Pull request merged**
